File: c++/81. Search in Rotated Sorted Array II.cpp
class Solution {
public:
	bool search(vector<int>& nums, int target) {
		if (nums.empty()) return false;
		int i;
		for (i = 0; i < nums.size() - 1; i++) {
			if (nums[i] > nums[i + 1])
				break;
		}
		int l, r, len = nums.size();
		if (target > nums[i]) return false;
		if (i + 1 < len && target < nums[i + 1]) return false;
		if (target <= nums[i] && target >= nums[0]) {
			l = 0; r = i;
		}
		else if (target >= nums[i + 1] && target <= nums[len - 1]) {
			l = i + 1; r = len - 1;
		}
	
		while (l <= r) {
			int mid = (l + r) >> 1;
			if (nums[mid] == target)
				return true;
			else if (nums[mid] > target)
				r = mid - 1;
			else
				l = mid + 1;
		}

		return false;
	}
};

File: go/src/leetcode/81. Search in Rotated Sorted Array II/81. Search in Rotated Sorted Array II.go
package main

import "fmt"

func search(nums []int, target int) bool {
	n := len(nums)
	l, r := 0, n-1
	var mid int
	for l <= r {
		for l < r && nums[l] == nums[l+1] {
			l++
		}
		for l < r && nums[r] == nums[r-1] {
			r--
		}

		mid = l + (r - l)/2
		k := nums[mid]
		if k == target {
			return true
		}
		if k < nums[r] { // mid-r 有序
			if k < target && target <= nums[r] {
				l = mid + 1
			} else {
				r = mid - 1
			}
		} else { // l-mid 有序
			if k > target && target >= nums[l] {
				r = mid - 1
			} else {
				l = mid + 1
			}
		}
	}

	return false
}

func main() {
	fmt.Println(search([]int{2,5,6,0,0,1,2}, 0))
	fmt.Println(search([]int{2,5,6,0,0,1,2}, 3))
}
